sha2/src/sha512/aarch64.rs

```diff
@@ -0,0 +1,210 @@
+    let mut s3 = vreinterpretq_u64_u8(vrev64q_u8(vld1q_u8(block[48..64].as_ptr())));
+    let mut s4 = vreinterpretq_u64_u8(vrev64q_u8(vld1q_u8(block[64..80].as_ptr())));
+    let mut s5 = vreinterpretq_u64_u8(vrev64q_u8(vld1q_u8(block[80..96].as_ptr())));
+    let mut s6 = vreinterpretq_u64_u8(vrev64q_u8(vld1q_u8(block[96..112].as_ptr())));
+    let mut s7 = vreinterpretq_u64_u8(vrev64q_u8(vld1q_u8(block[112..128].as_ptr())));
+
+    // Rounds 0 and 1
+    let mut initial_sum = vaddq_u64(s0, vld1q_u64(&K64[0]));
+    let mut sum = vaddq_u64(vextq_u64(initial_sum, initial_sum, 1), gh);
+    let mut intermed = vsha512hq_u64!(sum, vextq_u64(ef, gh, 1), vextq_u64(cd, ef, 1));
+    gh = vsha512h2q_u64!(intermed, cd, ab);
+    cd = vaddq_u64(cd, intermed);
+
+    // Rounds 2 and 3
+    initial_sum = vaddq_u64(s1, vld1q_u64(&K64[2]));
+    sum = vaddq_u64(vextq_u64(initial_sum, initial_sum, 1), ef);
+    intermed = vsha512hq_u64!(sum, vextq_u64(cd, ef, 1), vextq_u64(ab, cd, 1));
+    ef = vsha512h2q_u64!(intermed, ab, gh);
+    ab = vaddq_u64(ab, intermed);
+
+    // Rounds 4 and 5
+    initial_sum = vaddq_u64(s2, vld1q_u64(&K64[4]));
+    sum = vaddq_u64(vextq_u64(initial_sum, initial_sum, 1), cd);
+    intermed = vsha512hq_u64!(sum, vextq_u64(ab, cd, 1), vextq_u64(gh, ab, 1));
+    cd = vsha512h2q_u64!(intermed, gh, ef);
+    gh = vaddq_u64(gh, intermed);
+
+    // Rounds 6 and 7
+    initial_sum = vaddq_u64(s3, vld1q_u64(&K64[6]));
+    sum = vaddq_u64(vextq_u64(initial_sum, initial_sum, 1), ab);
+    intermed = vsha512hq_u64!(sum, vextq_u64(gh, ab, 1), vextq_u64(ef, gh, 1));
+    ab = vsha512h2q_u64!(intermed, ef, cd);
+    ef = vaddq_u64(ef, intermed);
+
+    // Rounds 8 and 9
+    initial_sum = vaddq_u64(s4, vld1q_u64(&K64[8]));
+    sum = vaddq_u64(vextq_u64(initial_sum, initial_sum, 1), gh);
+    intermed = vsha512hq_u64!(sum, vextq_u64(ef, gh, 1), vextq_u64(cd, ef, 1));
+    gh = vsha512h2q_u64!(intermed, cd, ab);
+    cd = vaddq_u64(cd, intermed);
+
+    // Rounds 10 and 11
+    initial_sum = vaddq_u64(s5, vld1q_u64(&K64[10]));
+    sum = vaddq_u64(vextq_u64(initial_sum, initial_sum, 1), ef);
+    intermed = vsha512hq_u64!(sum, vextq_u64(cd, ef, 1), vextq_u64(ab, cd, 1));
+    ef = vsha512h2q_u64!(intermed, ab, gh);
+    ab = vaddq_u64(ab, intermed);
+
+    // Rounds 12 and 13
+    initial_sum = vaddq_u64(s6, vld1q_u64(&K64[12]));
+    sum = vaddq_u64(vextq_u64(initial_sum, initial_sum, 1), cd);
+    intermed = vsha512hq_u64!(sum, vextq_u64(ab, cd, 1), vextq_u64(gh, ab, 1));
+    cd = vsha512h2q_u64!(intermed, gh, ef);
+    gh = vaddq_u64(gh, intermed);
+
+    // Rounds 14 and 15
+    initial_sum = vaddq_u64(s7, vld1q_u64(&K64[14]));
+    sum = vaddq_u64(vextq_u64(initial_sum, initial_sum, 1), ab);
+    intermed = vsha512hq_u64!(sum, vextq_u64(gh, ab, 1), vextq_u64(ef, gh, 1));
+    ab = vsha512h2q_u64!(intermed, ef, cd);
+    ef = vaddq_u64(ef, intermed);
+
+    for t in (16..80).step_by(16) {
```

Comment:
Maybe it's worth to fully unroll this loop?

Reply:
It doesn’t seem like it.

Before:

```
test sha512_10    ... bench:           6 ns/iter (+/- 0) = 1666 MB/s
test sha512_100   ... bench:          65 ns/iter (+/- 0) = 1538 MB/s
test sha512_1000  ... bench:         637 ns/iter (+/- 11) = 1569 MB/s
test sha512_10000 ... bench:       6,308 ns/iter (+/- 80) = 1585 MB/s
```

After:

```
test sha512_10    ... bench:           6 ns/iter (+/- 0) = 1666 MB/s
test sha512_100   ... bench:          64 ns/iter (+/- 3) = 1562 MB/s
test sha512_1000  ... bench:         639 ns/iter (+/- 58) = 1564 MB/s
test sha512_10000 ... bench:       6,374 ns/iter (+/- 765) = 1568 MB/s
```